### PR TITLE
[#583][Sotw] Ensure resources are properly sent again if envoy unsubscribes then subscribes again to a resource

### PR DESCRIPTION
**Pull request by valerian-roche**
This PR is part of the fix for the issue related to resources being unsubscribed from then subscribed to again
This one applies to both simple and linear cache when using the sotw protocol.
Actionable cases would likely only impact sotw-ADS as EDS outside of ADS does not share streams and will therefore usually not have multiple resources this way

This PR is built on top of https://github.com/envoyproxy/go-control-plane/pull/584 as it's relying on the client state being provided to the cache

It is also fixing a few bugs found during testing:
- in linear cache, the request set in the response was not the actual request. It was breaking some sotw parts as this was actually relied upon for some behavior (though not the bug of the issue)
- in linear cache, when using sotw, if a request has multiple resources (at least three) and receive at least two subsequent updates of subsets of those resources, there is a deadlock as the watch is leaked in the watches of resources requested but not yet updated. As this channel won't be listened after the first reply, the second update will fill the buffer and any subsequent cache update impacting a resource requested and not yet updated will deadlock

It is updating a few other aspects in order to implement the fix:
- Adds a staged resources map within the streamState. In sotw it is storing resources returned but not yet ACKed. This behavior should be ported to delta, but is out of the scope here. It allows removing the bespoke lastResponse value set only in sotw and parallel to streamState
- Adds a new returned attribute in the rawResponse, providing the list of resources included in the opaque response. This is needed to know which resources are actually known to the client. The previous implementation based on the request was actually opening edge cases as there is no guarantee a requested resource has ever been replied
- Adds cache logging in linear_test.go (helping while writing tests)

Finally, reasoning on the fixing through proper tracking of resources rather than basic resend of everything on doubts is explained within #584, and is even more applicable in the context of this issue (EDS through sotw-ads) when a single client can track thousands of endpoints within hundreds of CLAs

### CODE REVIEW

File: pkg/server/sotw/v3/server.go [hunk 1]
```diff
@@ -183,35 +164,53 @@ func (s *server) process(str stream.Stream, reqCh <-chan *discovery.DiscoveryReq
...
-				// If these requirements aren't satisfied, leave an open watch.
-				if w.nonce == "" || w.nonce == nonce {
-					w.close()
-
-					watches.addWatch(typeURL, &watch{
-						cancel:   s.cache.CreateWatch(req, streamState, responder),
-						response: responder,
-					})
+				// If we had an open watch, close it to make sure we don't end up sending a cache response while we update the state of the request
+				w.close()
```
Comment (valerian-roche) [inline pkg/server/sotw/v3/server.go:186]: The race mentioned here is currently present in the code if we receive a new request while we are building a response in the cache (race on KnownResourceNames)

**Comment by mattconzen** [comment]
@valerian-roche Is this one still needed in upstream or should it remain stale/closed?

**Comment by valerian-roche** [comment]
> @valerian-roche Is this one still needed in upstream or should it remain stale/closed?

Those PRs are addressing issues which are still present. There has been no feedback on it or https://github.com/envoyproxy/go-control-plane/pull/584 which is a pre-requisite of those PRs
Maybe @alecholmez can provide some feedback on when it could be reviewed

File: pkg/server/sotw/v3/server.go [hunk 2]
```diff
@@ -227,7 +226,24 @@ func (s *server) process(str stream.Stream, reqCh <-chan *discovery.DiscoveryReq
 				return err
 			}
 
-			watches.responders[res.GetRequest().TypeUrl].nonce = nonce
+			// Track the resources returned in the response
+			// Those are staged pending the client ACK
+			// The protocol clearly states that if we send another response prior to an ACK
+			// the previous one is to be considered as discarded
```
Comment (atollena) [inline pkg/server/sotw/v3/server.go:232]: Can you link to the protocol section that states that? I don't find the wording and I'm not sure this concern is valid.

Not having to resend resources that were already sent would make the code much simpler, as you would not need to track aknowledgement, afaik (no concept of "pending" resource).

**Comment by atollena** [comment]
Thanks for working on this, this is great.

This would fix https://github.com/envoyproxy/go-control-plane/issues/431 which makes go-control-plane pretty much incompatible with grpc clients, as those typically unsubscribe to resources after a channel reaches its idle timeout.

File: pkg/cache/v3/cache.go [hunk 3]
```diff
@@ -37,6 +36,27 @@ type Request = discovery.DiscoveryRequest
...
+// ClientState provides additional data on the client knowledge for the type matching the request
+// This allows proper implementation of stateful aspects of the protocol (e.g. returning only some updated resources)
+// Though the methods may return mutable parts of the state for performance reasons,
+// the cache is expected to consider this state as immutable and thread safe between a watch creation and its cancellation
+type ClientState interface {
+	// GetKnownResources returns the list of resources the clients has ACKed and their associated version.
+	// The versions are:
+	//  - delta protocol: version of the specific resource set in the response
+	//  - sotw protocol: version of the global response when the resource was last ACKed
+	GetKnownResources() map[string]string
```
[the remark this answers is not in the document]
Comment (valerian-roche) [reply pkg/cache/v3/cache.go:48]: Updated to this name. I have not pushed this branch yet, but I updated #584

File: pkg/cache/v3/cache.go [hunk 4]
```diff
@@ -37,6 +36,27 @@ type Request = discovery.DiscoveryRequest
 // DeltaRequest is an alias for the delta discovery request type.
 type DeltaRequest = discovery.DeltaDiscoveryRequest
 
+// ClientState provides additional data on the client knowledge for the type matching the request
+// This allows proper implementation of stateful aspects of the protocol (e.g. returning only some updated resources)
+// Though the methods may return mutable parts of the state for performance reasons,
+// the cache is expected to consider this state as immutable and thread safe between a watch creation and its cancellation
```
[the remark this answers is not in the document]
Comment (valerian-roche) [reply pkg/cache/v3/cache.go:42]: Updated on parent PR #584, not yet rebased on this branch

File: pkg/cache/v3/linear.go [hunk 5]
```diff
@@ -324,45 +335,74 @@ func (cache *LinearCache) CreateWatch(request *Request, streamState stream.Strea
...
+			cache.log.Debugf("Watch is stale as version failed to parse %s", err.Error())
+		}
+	} else if clientState.IsWildcard() {
 		stale = lastVersion != cache.version
+		if cache.log != nil {
+			cache.log.Debugf("Watch is stale as version differs for wildcard watch")
+		}
 	} else {
+		// Non wildcard case, we only reply resources that have effectively changed since the version set in the request
+		// This is used for instance in EDS
```
[the remark this answers is not in the document]
Comment (valerian-roche) [reply pkg/cache/v3/linear.go:348]: #540 mentions this issue in the case of Sotw-ads on linear cache for xds-grpc

File: pkg/server/sotw/v3/server.go [hunk 6]
```diff
@@ -183,35 +164,53 @@ func (s *server) process(str stream.Stream, reqCh <-chan *discovery.DiscoveryReq
...
+			// State cannot be modified until any potential watch is closed
+			state, ok := streamStates[typeURL]
+			if !ok {
+				// We don't have a current state for this type, create one
+				state = stream.NewStreamState(len(req.ResourceNames) == 0, nil)
+			} else if nonce != state.LastResponseNonce() {
+				// The request does not match the last response we sent.
+				// The protocol is a bit unclear in this case, but currently we discard such request and wait
+				// for the client to acknowledge the previous response.
+				// This is unclear how this handles cases where a response would be missed as any subsequent request will be discarded.
```
[the remark this answers is not in the document]
Comment (atollena) [reply pkg/server/sotw/v3/server.go:177]: The nonce is actually per type iiuc? That could be documented more clearly... The race that it exists to prevent only involve individual resource type. I think making the nonce per ADS stream (cross resource) would not work well: clients may not receive a newer discovery response but still be considered stale.

There is actually one case to at least take into consideration, explained here: https://github.com/envoyproxy/envoy/issues/10363.

File: pkg/server/sotw/v3/server.go [hunk 7]
```diff
@@ -183,35 +164,53 @@ func (s *server) process(str stream.Stream, reqCh <-chan *discovery.DiscoveryReq
...
-				// We've found a pre-existing watch, lets check and update if needed.
-				// If these requirements aren't satisfied, leave an open watch.
-				if w.nonce == "" || w.nonce == nonce {
-					w.close()
-
-					watches.addWatch(typeURL, &watch{
-						cancel:   s.cache.CreateWatch(req, streamState, responder),
-						response: responder,
-					})
+				// If we had an open watch, close it to make sure we don't end up sending a cache response while we update the state of the request
```
[the remark this answers is not in the document]
Comment (atollena) [reply pkg/server/sotw/v3/server.go:185]: Your response is clearer.

File: pkg/server/sotw/v3/server.go [hunk 8]
```diff
@@ -183,35 +164,53 @@ func (s *server) process(str stream.Stream, reqCh <-chan *discovery.DiscoveryReq
...
+				w.close()
+
+				// Check if the new request ACKs the previous response
+				// This is defined as nonce and versions are matching, as well as no error present
+				if state.LastResponseNonce() == nonce && state.LastResponseVersion() == req.VersionInfo && req.ErrorDetail == nil {
```
[the remark this answers is not in the document]
Comment (atollena) [reply pkg/server/sotw/v3/server.go:190]: Yes, so the list of subscribed revisions would need to be part of the version string.

> Envoy requires to send a response even if the version does not change for some phases of load

OK, so I guess my hope that we could avoid sending a response on stream open if it is already up to date is not something we can support, even in theory. Anyways, this is not the focus of this PR so I'll close this comment for now.

File: pkg/server/sotw/v3/server.go (hunk 2 above)
Comment (atollena) [reply pkg/server/sotw/v3/server.go:232]: OK, could you validate envoy's behavior in the case you describe? gRPC will cache versions sent at version N, and only update those that changed at N+1, so nothing is lost. I think it'd be tricky to even implement something different on the client side, but perhaps Envoy does.

**Comment by carsonoid** [comment]
> Thanks for working on this, this is great.
> 
> This would fix #431 which makes go-control-plane pretty much incompatible with grpc clients, as those typically unsubscribe to resources after a channel reaches its idle timeout.

@atollena the patch [here](https://github.com/envoyproxy/go-control-plane/issues/431#issuecomment-2021010775) fixes the go client idle issue for me. Hopefully that it can help you too

**Comment by dfawley** [comment]
What is the status of this PR?  What would it take to make it finally land?  We just got bit by this again in some proposed testing we were going to add to gRPC to catch races creating and destroying gRPC channels using xDS, but can't until this is fixed.

**Comment by valerian-roche** [comment]
> What is the status of this PR? What would it take to make it finally land? We just got bit by this again in some proposed testing we were going to add to gRPC to catch races creating and destroying gRPC channels using xDS, but can't until this is fixed.

Hey Doug,
Within my company we have been using this setup for a while with grpc xDS clients (actually using the branch https://github.com/DataDog/go-control-plane/tree/dd/main). 
This branch relies on an API breaking change in the Cache interface which, for most users, should not have any impact, but creates an issue with the single tag release we used to have.
Once the dust has settled on the module split I hope to be able to push those changes on the upstream main branch.
